js/pinnedLinks.js

```diff
@@ -0,0 +1,142 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+ /*globals gUserDatabase, OBJECT_STORE_PREFS, PINNED_LINKS_PREF */
+
+"use strict";
+
+/**
+ * Singleton that keeps track of all pinned links and their positions in the
+ * grid.
+ */
+
+(function(exports) {
+  const gPinnedLinks = {
+    /**
+     * The cached list of pinned links.
+     */
+    _links: [],
```

Comment:
The way it's currently implemented, `_links` is intended to be an ordered list so I don't think it will directly translate to a Set or Map. Mind you, I'm not saying this is the best way to represent `_links` but there's a lot of design that we're porting over from Firefox that perhaps we can re-think once we've got it in a working state.